File: bin/m_toss.py
#!/usr/bin/env python
import os
import subprocess
import re

import read_data_pm as readpm

class Param():
    def __init__(self, cname=None, appli=None, qclass=None, cores=None, identifier=None, cpns=None, time=None):
        self.cname = cname
        self.appli = appli
        self.qclass = qclass
        self.cores = cores
        if identifier != 'jss':
            self.identifier = identifier
        self.cpns = cpns
        self.time = self.time

    def __str__(self):
        return '.'.join(filter(None([self.cname, self.appli, self.qclass, self.cores, self.identifier, 'jss'])))

    def __repr__(self):
        return '.'.join([self.cname, self.appli, self.qclass, self.cores, self.identifier, self.cpns, self.time])

    def __eq__(self, other):
        if isinstance(other, Param):
            # how to define this array of applis
            # just use string compare in (ugly I know)
            if self.qclass == other.qclass and self.cores == other.cores and self.identifier == other.identifier
            ( self.appli in other.appli or other.appli in self.apli):
                return True
        return False

def create_jss(requested_parameter, jss_dir=os.path.join(os.environ['CLUSTER_SCRIPTS_TOP'], 'jobSubScripts')):
    def table_lookup(filename, parameter):
        with open(filename, 'r') as data_fh:
            lines = data_fh.readlines()

            # get the keys
            # because 'class' is a python keyword, change class -> qclass
            keys = [ x.lower() for x in lines[0].split() if x.lower() != 'class' else 'qclass' ]

            # also skip '------' (lines[1])
            data_lines = lines[2:]

            for line in data_lines:
                cols =  [ l.lower() for l in line.split() ]
                if not cols:
                    continue
                # don't forget to check if zipping identifier to none
                available_parameter = Param(**{ key.lower(): value for key, value in zip(keys, line.split()) })
                print(repr(available_parameter)
                if available_parameter == parameter:
                    available_parameter.appli = parameter.appli
                    return available_parameter

            # TODO: if not found, read qc and ask if user want to make a new one
            # basically see if the input value is within the range on qc
            # qc = readpm.read_perl_module_hashes('/mnt/lustre/bin2/qc')
            raise Exception
        #endwith
    #enddef

    def construct_jss(jss_template, parameter):
        with open(jss_template, 'r') as template_fh:
            jss = template_fh.read()
            if '#VERSION' in content:
                version = readpm.input_until_correct(
                    'The available versions are {}\n'.format(' '.join(line.split()[1:])),
                    lambda x: x in line.split()[1:])
                jss = jss.replace('VERSION', version)
            for key in vars(parameter).keys():
                jss.replace(key.upper(), parameter.key.upper())
            jss.replace('JOBNAME', jname)
            return jss
        #endwith
    #enddef

    config_table = os.path.join(jss_dir, '{}.data'.format(requested_parameter.cname))
    jss_template = os.path.join(jss_dir, '{}.{}'.format(requested_parameter.cname, requested_parameter.appli))

    print(str(requested_parameter))

    final_parameter = table_lookup(config_table, requested_parameter)
    sub_filename = str(final_parameter)
    exit()

    # can this be done without side-effect (instead of writing here, return string to be written as file) ?
    with open(sub_filename, 'w') as sub_fh:
        sub_fh.write(construct_jss(jss_template, final_parameter))
    return sub_filename

# functions related to the k-computer? I don't know if these are still relevant
def unlink(link):
    print(' '.join(['unlink', link]))
    #subprocess.call(['unlink', link])
def link(filename, link):
    print(' '.join(['ln', '-s', filename, link]))
    #subprocess.call(['ln', '-s', filename, link])

if __name__ == '__main__':
    # parse arguments
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('param',
            help='Toss parameter (see chk.pl), in this format: MACHINE.APPLI.CLASS.CORES(.IDENITFIER) or MACHINE.(anything).jssc',
            metavar='SPEC')
    parser.add_argument('--make-jobdir', action='store_true')
    parser.add_argument('--no-delete', action='store_true')
    args, rsync_args = parser.parse_known_args()
    tosssyncopt = '--delete' if not args.no_delete else ''

    # read perl module
    datapm = readpm.read_perl_module_hashes('/mnt/lustre/scriptsForClusterNew/data.pm')

    uname = readpm.get_uname()
    jname = readpm.get_jname()

    # TODO: comparing with template is perhaps better?
    # rule 1: always re-submit modified files regardless what;
    # rule 2: do not delete modified job scripts
    # pseudo code:
    # make temp_job_script
    # if parameter is a file and exist and different from temp
    #   toss with file; rm temp
    # else
    #   toss using temp
    #   delete previous modifications
    # deletion should only be done if wanted
    # subprocess.call(['rm', '*.jss'])
    # print('rm *.jss')

    parameter = Param(*[ s.lower()  for s in args.param.split('.') ])

    job_script = args.param if re.match(r'[A-Za-z0-9]*.jssc', args.param) else create_jss(parameter)
    exit()

    # it's here because we have to define the cluster name first
    cname = readpm.specify_cluster()

    # k supercomputer stuffs, are these relevant?
    if 'mkei' in cname:
        unlink('casino')
        unlink('shm_casino')
        unlink('espresso')
        link('ln -s /home/hp150271/k03110/applications/v213pl662/bin_qmc/kei/opt/casino', 'casino')
        link('ln -s /home/hp150271/k03110/applications/v213pl662/bin_qmc/kei/Shm/opt/casino', 'shm_casino')
    elif 'ekei' in cname:
        unlink('casino')
        unlink('shm_casino')
        unlink('espresso')
        link('ln -s /home/hp160251/k03241/applications/v213pl662/bin_qmc/kei/opt/casino', 'casino')
        link('ln -s /home/hp160251/k03241/applications/v213pl662/bin_qmc/kei/Shm/opt/casino', 'shm_casino')

    # data.pm data
    c1 = datapm['clusters']['unified']
    c2 = datapm['clusters'][cname]
    print(c1['pbs_dir'])

    # related to JobDir file
    # TODO: perhaps we can do without calling? need to restructure m_jobdir
    if args.make_jobdir:
        subprocess.call('m_jobdir.pl')

    # put the files
    put = [ os.path.join(os.environ['CLUSTER_SCRIPTS_TOP'], 'm_put.pl'), '-c', cname, '--dest', path, tosssyncopt ]
    put.extend(rsync_args)
    print(subprocess.list2cmdline(put))
    #subprocess.call(put)

    # qsub and save the job number
    get_jobnumber = [ 'ssh', c2['ssh_option'], readpm.c2ip(c2, uname), '\"cd {} ; {} {}\"'.format(path, os,path.join(c2['pbs_dir'], c2['submit']), job_script) ]
    print(subprocess.list2cmdline(get_jobnumber))
    with open('JobNumber', 'w') as jobnumber:
        lines = readpm.get_command_outputs(ret_jn_cmd, readpm.c2ip(c2, uname))
        assert len(lines) == 1
        jobnumber.write(lines[0])

    # don't forget to put it too
    put_jobnumber = [ put, 'JobNumber' ]
    print(subprocess.list2cmdline(put_jobnumber))
    #subprocess.call(put_jobnumber)

File: bin/m_toss.pl
#!/usr/bin/perl
# ############################################
#
#   14Mar19  ;  Allow to manually specify jobname as argument (postponed)
#   25Oct18  ;  Implement dry run and jobdir option
#   14Jun18  ;  Fix indentations and general readability in old part of the code
#   25Apr18  ;  Tweaks and variable renaming (depart from previous confusing scheme)
#               Now read JobDir if available
#   10Apr18  ;  remove enhanced '-e' option
#   02Apr18  ;  add enhanced '-e' option
#               Append 'touch COMPLETED' at the end of sub file when -e option is added
#               remove 'COMPLETED' file at toss when -e option is used
#   14Feb18  ;  Put target directory to a (?) file
#   05Feb18  ;  Fix shown script name
#   03Feb18  ;  Add option to not delete the files (for continuing partial-fetcheds)
#   30Jan18  ;  Designate new extension '.jssc' for custom jss, to allow different handling of it
#               New format: MACHINE.(anything you want).jssc
#   29Jan18  ;  Prevent deletion of other jss while creating jss (surprisingly BAD behaviour here)
#   26Jan18  ;  Prevent deleting jss when incorrectly inputting arguments
#   12Jan18  ;  Forked by genki
#   23Jan18  ;  Do not delete jss when no matching MACHINE/.../IDENTIFIER found
#               Added functionality to skip jss making,
#               allowing the use of custom jss (by using jss file as argument)
#               Note that for compatibility with the normal workflow using other scripts,
#               jss file name should be in the format of MACHINE.(anything you want).jss
#   24Aug15  ;  First composed by ichibha
#
############################################

use Cwd;
use lib $ENV{CLUSTER_SCRIPTS_TOP};
use data;
use Getopt::Long;# qw(:config permute);
use File::Basename;
#use POSIX qw(SIGINT);
$0=basename($0);

my $a = '';
# TODO: generalize
#sub processArg {
#    $a = $_;
#    @w = split '\.', $a;
#}
GetOptions(
    'no-delete'   => sub { $no_delete = 1 },
    'dry-run'     => sub { $dry_run   = 1 },
    'jobdir' => sub { $make_jobdir = 1 },
) or die "$0: Invalid arguments\n";
$tosssyncopt = $no_delete ? '' : '--delete';
#    'name' => \my $jname,
#    '<>' => \&processArg,

my $a = $ARGV[0];
if($a eq ''){
    print "USAGE: $0 MACHINE.APPLI.CLASS.CORES(.IDENITFIER) or MACHINE.(anything).jssc\n";
    exit 0;
}



### making jss ###
&data::getUserName();
$uname = $data::userName;
&data::getJobName() ;
$jname = defined $jname? $jname: $data::jobName ;
open UNAME, '> UserName';
print UNAME "$uname\n";
close UNAME;

@w = split '\.', $a;
$cname = $w[0];
$appli = $w[1];
$class = $w[2];
$cores = $w[3];
$identifier = $w[4];
if ($a =~ /.jssc/) { #Use argument as -custom- job script whenever exists
    die "$0: $a: No such file\n" unless -e $a;
    $jssfile = $a;
    chomp $jssfile;
} else {
    if ($identifier =~ /jss/) {
        $identifier = '';
    }

    $dataf = $ENV{CLUSTER_SCRIPTS_TOP}.'/jobSubScripts/'.$cname.'.data';
    die "$0: $dataf: No such file\n" unless -e $dataf;
    $tempf = $ENV{CLUSTER_SCRIPTS_TOP}.'/jobSubScripts/'.$cname.'.'.$appli;
    die "$0: $tempf: No such file\n" unless -e $tempf;

    open  DATA ,"< $dataf";
    $_ = <DATA>;
    @keys = split;
    while (<DATA>) {
        chomp;
        @w = split;
        if ( $w[0] =~ /$class/i && $w[1] == $cores
    	  && $w[$#keys] eq $identifier
          && $class ne '' && $cores ne '' && $appli ne ''
    	  && $w[$#keys-1] =~ /$appli/i ) {
    	    for ($i=0;$i<=$#w;$i++) {
    	        $para{$keys[$i]} = $w[$i];
    	    }
    	    last;
        }
        die "$0: No matching MACHINE/APPLI/CLASS/CORES/IDENITFIER found (check with chk.pl)\n" if eof;
    }
    close DATA;

    system("rm *.jss\n");
    if ($para{IDENTIFIER} eq '') {
        $jssfile = "$cname\.$appli\.$para{CLASS}\.$para{CORES}\.jss";
    } else {
        $jssfile = "$cname\.$appli\.$para{CLASS}\.$para{CORES}\.$para{IDENTIFIER}\.jss";
    }

    open TEMP, "< $tempf";
    open JSS,  "> $jssfile";
    while (<TEMP>) {
        if (/\A#VERSION/) {
    	    print JSS $_;
    	    chomp;
    	    @w = split;
    	    print "The available versions are";
    	    for ($i=1;$i<=$#w;$i++) {
    	        print " $w[$i]";
    	    }
    	    print "\n";
    	    $version = `cat version`;
    	    chomp $version;
    	    if ($version ne '') {
    	        print "The version $version is detected\n";
    	    } else {
    	        print "please input the version\n";
    	        $version =  <STDIN>;
    	        chomp $version;
    	    }
    	    for ($i=1;$i<=$#w;$i++) {
    	        if ($version == $w[$i]) {
    	    	    print "Ver. $version is available.\n";
    	    	    last;
    	        }
    	        if ($i == $#w) {
    	    	    die "Ver. $version is not found.";
    	        }
    	    }
    	    system "echo $version > version";
    	    $_ = <TEMP>;
    	    $_ =~ s/VERSION/$version/g;
        }
        foreach $key (@keys){
    	    $_ =~ s/$key/$para{$key}/g;
        }
            $_ =~ s/JOBNAME/$jname/;
        print JSS $_;
    }
    close TEMP;
    close JSS;
}
### version ###



##### qsub part
if($cname =~ /mkei/) {
    system 'unlink casino';
    system 'unlink shm_casino';
    system 'unlink espresso';
    system 'ln -s /home/hp150271/k03110/applications/v213pl662/bin_qmc/kei/opt/casino casino';
    system 'ln -s /home/hp150271/k03110/applications/v213pl662/bin_qmc/kei/Shm/opt/casino shm_casino';
} elsif ($cname =~ /ekei/) {
    system 'unlink casino';
    system 'unlink shm_casino';
    system 'unlink espresso';
    system 'ln -s /home/hp160251/k03241/applications/v213pl662/bin_qmc/kei/opt/casino casino';
    system 'ln -s /home/hp160251/k03241/applications/v213pl662/bin_qmc/kei/Shm/opt/casino shm_casino';
}

# data.pm data
my $c1dir = $data::clusters{unified}{base_dir};
my $c2dir = $data::clusters{$cname}{base_dir};
my $c2ip   = $data::clusters{$cname}{ip_address}; $c2ip   =~ s/USER/$uname/g;
my $c2pbs  = $data::clusters{$cname}{pbs_dir};
my $c2sub  = $data::clusters{$cname}{submit};
my $c2sshopt = $data::clusters{$cname}{ssh_option};
my %cluster = %{$data::clusters{$cname}};
print %cluster{'pbs_dir'};
my $cwd = Cwd::getcwd().'/';

# related to JobDir file
my $path;
if (-s 'JobDir') {
    open JOBDIR, "< JobDir";
    $path = <JOBDIR>; chomp $path;
    print "Path in JobDir: $path\n";
    close JOBDIR;
} else {
    my $base = $cwd;
    $base =~ s/$c1dir//;
    $path = "$c2dir$base";
    $path =~ s/USER/$uname/;
    if ($make_jobdir && not $dry_run) {
        open JOBDIR, "> JobDir";
        print JOBDIR $path."\n";
        close JOBDIR;
    }
}

# if dry run, do nothing
if ($dry_run) {
    my $put = "$ENV{CLUSTER_SCRIPTS_TOP}/m_put --dry-run -c $cname --dest $path";
    system "$put $tosssyncopt";
} else {
    my $put = "$ENV{CLUSTER_SCRIPTS_TOP}/m_put -c $cname --dest $path";
    system "$put $tosssyncopt";
    #my $jobnum = `ssh $c2sshopt $c2ip "cd $path ; $c2pbs$c2sub -N $jname $jssfile"`; #pending
    my $jobnum = `ssh $c2sshopt $c2ip "cd $path ; $c2pbs$c2sub $jssfile"`;
    open JOBNUM, "> JobNumber";
    print JOBNUM $jobnum;
    close JOBNUM;
    if (-s 'JobDir') {
        system "$put JobNumber JobDir";
    } else {
        system "$put JobNumber";
    }
}
